### ISSUE TITLE
TetGen 1.6.0

### ISSUE THREAD

**Issue by mastep84** [UG4/tool_tetgen#2]
Dear all,

TetGen version 1.6.0 has been released on August 31, 2020 which once again introduces a substantial boost to the quality of the generated tetrahedralizations (see http://wias-berlin.de/software/tetgen/relnotes.html). Are there any objections from your side to updating tool_tetgen (currently featuring the previous version 1.5.0) accordingly?

Best,
MS

**Comment by sreiter** [comment]
Hey Martin,

Thank's for pointing that out. Since ProMesh uses the tetgen tool, too, I guess it would make sense to copy the tetgen 1.6 executable to the ProMesh4/tools directory and test whether tetrahedralizations still work as expected, before updating to the new version.

Regards,
Sebastian

**Comment by stephanmg** [comment]
That would be great to test it before! Out of curiosity: How do you test this? Are there any unit tests?

**Comment by mastep84** [comment]
Hi Sebastian and Stephan,

thank you for your feedback! I took the liberty and tested ProMesh 4.3.18 with the tetgen 1.6 executable in the ProMesh4/tools directory in place. As the command line switches for tetgen 1.6 weren't modified there are no issues to report here at least from the technical side. 

However, now the default algorithm for the -p option (boundary recovery) is the newly improved constrained tetrahedralization algorithm. It is claimed to be robust and use less Steiner points than the constrained Delaunay tetrahedralization (CDT) algorithm which was previously set as default. This change is accompanied by a new implementation of the constrained Delaunay refinement algorithm (-q option) which uses new Steiner point insertion schemes to remove badly-shaped elements.

Against this background, I did some grid analysis comparing the tetrahedrizations of tetgen 1.4.3, 1.5 and 1.6 in case of reconstructed cell nuclei that I used for the subdivision-based geometric multigrid method development. The comparison included the geometric evaluation of minimal dihedral angles and discretization associated quality ratios as well as a brief numerical evaluation of the GMG solver performance. While tetgen 1.4.3 and 1.6 both succeeded in building coarse tetrahedral grids of equally good quality which were well-suited as base level grids for GMG, I found that using tetgen 1.5 with the same settings resulted in up to 10 times finer tetrahedrizations that did not even match the quality of the other two. That geometric impression was furthermore confirmed by the convergence rates and iteration numbers resp. which I observed in case of a Poisson model problem and that were nearly twice the size compared to the simulations involving the grids produced by tetgen 1.4.3 and 1.6.

From my personal experience I have already preferred the tetrahedrizations by tetgen 1.4.3 over those of 1.5 in many applications in the past before. So at least in my opinion, tetgen 1.6 definitely is an improvement and step in the right direction worth upgrading to.

Regards,
Martin

**Comment by stephanmg** [comment]
So 1.4.3 and 1.6 are roughly equivalent in terms of grid quality?

**Comment by mastep84** [comment]
Yes, that is my impression. The overall quality of tetgen 1.6 tetrahedrizations seems to be superior to those generated by version 1.5. However, the issue turns out to be more complicated than that. I did some additional analysis testing for the occurrences of sliver elements and this again drew a significantly different picture to the positive impression from before. Interestingly, tetgen 1.6 was much more prone to including flat slivers than version 1.5. And from our standpoint of FE and FV discretizations this is a critical issue because even a single ill-shaped element can ruin the condition of the stiffness matrix. Therefore I would recommend to stick with tetgen 1.5 after all -- at least for now and preferably until we can gather more robust experience with the 1.6 tetrahedrizations in our daily applications.

**Comment by stephanmg** [comment]
That is very interesting @mastep84.

Out of curiosity: I need a reference for exactly this statement you made:

> And from our standpoint of FE and FV discretizations this is a critical issue because even a single ill-shaped element can ruin the condition of the stiffness matrix

Is there a standard reference for this? Should one cite _Shewchuk, J. "What is a good linear finite element? interpolation, conditioning, anisotropy, and quality measures (preprint)." University of California at Berkeley 73 (2002)_
